packages/core/src/browser/saveable-service.ts

```diff
@@ -222,7 +222,8 @@ export class SaveableService implements FrontendApplicationContribution {
         if (!Saveable.isDirty(widget)) {
             return false;
         }
-        if (this.autoSave !== 'off') {
+        const saveable = Saveable.get(widget);
+        if (this.autoSave !== 'off' && (!saveable || this.shouldAutoSave(widget, saveable))) {
```

Comment:
I don't know the specifics of `Saveable`. Looking at the code, what's weird to me is that we return `true` in `shouldSaveWidget` if there is no saveable... Shouldn't it look like this?

```suggestion
        const saveable = Saveable.get(widget);
        if (this.autoSave !== 'off' && saveable && this.shouldAutoSave(widget, saveable)) {
```

Note that the use case is still fixed with this suggestion, but I may overlook some other consequences

Reply:
`saveable` should always be defined or we would never get here - we wouldn't even be considering saving the widget if we hadn't discovered a saveable earlier in the process, but we don't pass it into this particular method - so the two are logically equivalent, so I'm happy to go either way.

Reply:
So maybe like this
```
        const saveable = Saveable.get(widget);
        if (!saveable) {
          throw new Error('saveable does not exist in widget which is checked for saving')
          
          //or

          // should never happen
          return false;

          //or

          console.warn('saveable does not exist in widget which is checked for saving');
          return false;
        }
          
        if (this.autoSave !== 'off' && this.shouldAutoSave(widget, saveable)) {
```

or

```
        const saveable = Saveable.get(widget);          
        // saveable must always exist here
        if (this.autoSave !== 'off' && this.shouldAutoSave(widget, saveable!)) {
```

Reply:
The trouble with each of your proposals is that they will throw errors if something unexpected happens, and I wanted to avoid that outcome, because that amounts to an API breakage that is only detectable at runtime, which is particularly frustrating for adopters. If we are genuinely concerned about the case where `Saveable.get` returns either nothing or something other than the expected `Saveable`, then we should break the API in a way that is tranparent in the type system and pass the saveable on which this method is dependent into it.

In my view, it's fairly unlikely that we will get anything other than the expected `Saveable` from the `Saveable.get` call, so my original solution, which leaves adopters in the position (perhaps) having to fix the bug for themselves if they've done something that leads to the saveable being `undefined`, seemed like a reasonable compromise between  fixing the bug for the normal case and leaving the detectable abnormal case of `undefined` `Saveable` unaffected, but if you'd prefer that we be sure that we're checking the correct `Saveable`, I'd be happy to implement it accordingly.

Reply:
I get that it's unnecessary to change behavior for adopters, at the same time it's weird that we explicitly check for an "impossible" state and then arbitrarily return `true` in `shouldSaveWidget` for something we don't even expect to be saveable.

To be pragmatic, let's go with the original suggestion of yours but add a comment? e.g.

```suggestion
        const saveable = Saveable.get(widget);
        // saveable should always exist here. We enter the if branch if it doesn't to keep backward compatibility
        if (this.autoSave !== 'off' && (!saveable || this.shouldAutoSave(widget, saveable))) {
```